presto-main/src/main/java/com/facebook/presto/operator/SpatialIndexBuilderOperator.java

```diff
@@ -0,0 +1,207 @@
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("Spatial index build can not be duplicated");
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final LocalMemoryContext localUserMemoryContext;
+    private final PagesSpatialIndexFactory pagesSpatialIndexFactory;
+
+    private final List<Integer> outputChannels;
+    private final int indexChannel;
+    private final Optional<JoinFilterFunctionFactory> filterFunctionFactory;
+
+    private final PagesIndex index;
+
+    private boolean finished;
+
+    public SpatialIndexBuilderOperator(
+            OperatorContext operatorContext,
+            PagesSpatialIndexFactory pagesSpatialIndexFactory,
+            List<Integer> outputChannels,
+            int indexChannel,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory,
+            int expectedPositions,
+            PagesIndex.Factory pagesIndexFactory)
+    {
+        this.operatorContext = operatorContext;
+        this.localUserMemoryContext = operatorContext.localUserMemoryContext();
+        this.filterFunctionFactory = filterFunctionFactory;
+
+        this.index = pagesIndexFactory.newPagesIndex(pagesSpatialIndexFactory.getTypes(), expectedPositions);
+        this.pagesSpatialIndexFactory = pagesSpatialIndexFactory;
+
+        this.outputChannels = outputChannels;
+        this.indexChannel = indexChannel;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return pagesSpatialIndexFactory.getTypes();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+        checkState(!isFinished(), "Operator is already finished");
+
+        index.addPage(page);
+
+        if (!localUserMemoryContext.trySetBytes((index.getEstimatedSize().toBytes()))) {
+            index.compact();
+            localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes());
+        }
+
+        operatorContext.recordGeneratedOutput(page.getSizeInBytes(), page.getPositionCount());
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    @Override
+    public void finish()
+    {
+        if (finished) {
+            return;
+        }
+
+        PagesSpatialIndex spatialIndex = index.createPagesSpatialIndex(operatorContext.getSession(), indexChannel, filterFunctionFactory, outputChannels);
+        localUserMemoryContext.setBytes(spatialIndex.getEstimatedMemorySizeInBytes());
+        pagesSpatialIndexFactory.lendPagesSpatialIndex(spatialIndex);
+        finished = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finished;
```

Comment:
You cannot just finish here. Unless I am missing something, once you return `true` from `isFinished()`, the `Driver` can be finished too. `com.facebook.presto.operator.Driver#destroyIfNecessary` is going to cancel all memory allocations done by operators contained in this driver. Thus, this means the memory occupied by the `this.index` and `this. pagesSpatialIndexFactory.pagesSpatialIndex` will not be accounted for.

Reply:
@findepi Thanks for pointing this out. I updated the operators to make sure that builder operator doesn't finish until all probe operators have finished. I published these changes as a separate commit to simplify review: https://github.com/prestodb/presto/pull/9474/commits/5b3ccce90b5ead997d059496c25ae953ae3fc5c6

Please, take a look.